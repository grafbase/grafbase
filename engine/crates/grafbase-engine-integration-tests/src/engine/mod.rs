mod builder;

use std::{collections::HashMap, future::IntoFuture, sync::Arc};

use futures::{future::BoxFuture, Stream, StreamExt};
use grafbase_engine::{Request, RequestHeaders, Response, Schema, StreamingPayload, Variables};

pub use self::builder::EngineBuilder;

/// An instance of the grafbase-engine code that can be used for testing.
#[derive(Clone)]
pub struct Engine {
    inner: Arc<Inner>,
}

struct Inner {
    schema: Schema,
}

impl Engine {
    pub async fn new(schema: String) -> Self {
        EngineBuilder::new(schema).build().await
    }

    pub fn execute(&self, operation: impl Into<GraphQlRequest>) -> ExecutionRequest<'_> {
        ExecutionRequest {
            graphql: operation.into(),
            headers: HashMap::new(),
            schema: &self.inner.schema,
        }
    }

    pub fn execute_stream(&self, operation: impl Into<GraphQlRequest>) -> StreamExecutionRequest<'_> {
        StreamExecutionRequest {
            graphql: operation.into(),
            headers: HashMap::new(),
            schema: &self.inner.schema,
        }
    }
}

#[must_use]
pub struct ExecutionRequest<'a> {
    graphql: GraphQlRequest,
    headers: HashMap<String, String>,
    schema: &'a Schema,
}

impl ExecutionRequest<'_> {
    /// Adds a header into the request
    pub fn header(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.insert(name.into(), value.into());
        self
    }
}

impl<'a> IntoFuture for ExecutionRequest<'a> {
    type Output = Response;

    type IntoFuture = BoxFuture<'a, Response>;

    fn into_future(self) -> Self::IntoFuture {
        Box::pin(async move {
            let mut request = Request::new(self.graphql.query).data(RequestHeaders::from(&self.headers));
            if let Some(name) = self.graphql.operation_name {
                request = request.operation_name(name);
            }
            if let Some(variables) = self.graphql.variables {
                request = request.variables(variables);
            }
            self.schema.execute(request).await
        })
    }
}

#[must_use]
pub struct StreamExecutionRequest<'a> {
    graphql: GraphQlRequest,
    headers: HashMap<String, String>,
    schema: &'a Schema,
}

impl<'a> StreamExecutionRequest<'a> {
    /// Adds a header into the request
    pub fn header(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.insert(name.into(), value.into());
        self
    }

    /// Converts the execution request into a Stream
    pub fn into_stream(self) -> impl Stream<Item = StreamingPayload> + 'a {
        let mut request = Request::new(self.graphql.query).data(RequestHeaders::from(&self.headers));
        if let Some(name) = self.graphql.operation_name {
            request = request.operation_name(name);
        }
        if let Some(variables) = self.graphql.variables {
            request = request.variables(variables);
        }
        self.schema.execute_stream(request)
    }

    // Collects the StreamingPayloads into a vec
    pub async fn collect(self) -> Vec<StreamingPayload> {
        self.into_stream().collect().await
    }

    /// Converts the execution request into an iterator
    pub async fn into_iter(self) -> impl Iterator<Item = StreamingPayload> {
        self.collect().await.into_iter()
    }
}

pub struct GraphQlRequest {
    query: String,
    operation_name: Option<String>,
    variables: Option<Variables>,
}

impl From<&str> for GraphQlRequest {
    fn from(val: &str) -> Self {
        GraphQlRequest {
            query: val.into(),
            operation_name: None,
            variables: None,
        }
    }
}

impl From<String> for GraphQlRequest {
    fn from(val: String) -> Self {
        GraphQlRequest {
            query: val,
            operation_name: None,
            variables: None,
        }
    }
}

impl<T, V> From<cynic::Operation<T, V>> for GraphQlRequest
where
    V: serde::Serialize,
{
    fn from(operation: cynic::Operation<T, V>) -> Self {
        GraphQlRequest {
            query: operation.query,
            variables: Some(serde_json::from_value(serde_json::to_value(operation.variables).unwrap()).unwrap()),
            operation_name: operation.operation_name.map(|name| name.to_string()),
        }
    }
}
