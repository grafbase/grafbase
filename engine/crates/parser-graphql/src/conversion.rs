//! Allow converting a [`cynic_introspection::Schema`] into a new [`Registry`].
//!
//! The conversion takes all the known information from the introspection and adds it to the
//! registry.

use std::ops::Not;

use engine::{
    registry::{Deprecation, MetaDirective, MetaEnumValue, MetaField, MetaInputValue, MetaType, ObjectType, Registry},
    Value,
};

pub fn registry_from_introspection(schema: cynic_introspection::Schema) -> Registry {
    let mut registry = Registry::new();

    // root types
    registry.query_type = schema.query_type;
    registry.mutation_type = schema.mutation_type;
    registry.subscription_type = schema.subscription_type;

    // directives
    registry.directives = schema
        .directives
        .into_iter()
        .map(|d| (d.name.clone(), directive_from_introspection(d)))
        .collect();

    registry.implements = schema
        .types
        .iter()
        .filter_map(|ty| match ty {
            cynic_introspection::Type::Object(object) => object
                .interfaces
                .is_empty()
                .not()
                .then(|| (ty.name(), &object.interfaces)),
            cynic_introspection::Type::Interface(interface) => interface
                .interfaces
                .is_empty()
                .not()
                .then(|| (ty.name(), &interface.interfaces)),
            _ => None,
        })
        .map(|(name, interfaces)| (name.to_string(), interfaces.iter().cloned().collect()))
        .collect();

    // types
    registry.types = schema
        .types
        .into_iter()
        .map(|ty| (ty.name().to_owned(), type_from_introspection(ty)))
        .collect();

    registry
}

fn directive_from_introspection(directive: cynic_introspection::Directive) -> MetaDirective {
    MetaDirective {
        name: directive.name,
        description: directive.description,
        locations: directive
            .locations
            .into_iter()
            .map(directive_location_from_introspection)
            .collect(),
        args: directive
            .args
            .into_iter()
            .map(|v| (v.name.clone(), input_value_from_introspection(v)))
            .collect(),
        is_repeatable: false,
    }
}

fn field_from_introspection(field: cynic_introspection::Field) -> MetaField {
    MetaField {
        name: field.name,
        description: field.description,
        args: field
            .args
            .into_iter()
            .map(|v| (v.name.clone(), input_value_from_introspection(v)))
            .collect(),
        ty: field.ty.to_string().into(),
        deprecation: deprecated_from_introspection(field.deprecated),
        ..Default::default()
    }
}

fn input_value_from_introspection(input: cynic_introspection::InputValue) -> MetaInputValue {
    MetaInputValue {
        name: input.name,
        description: input.description,
        ty: input.ty.to_string().into(),
        default_value: input.default_value.and_then(convert_default_value),
        validators: None,
        is_secret: false,
        rename: None,
    }
}

fn convert_default_value(default_value: String) -> Option<Value> {
    // The default_value is a string in GraphQL format so we need to be careful to convert it correctly.
    // The main gotcha is that enum values such as `HELLO` should not be converted to string values
    // which would be represented as `"HELLO"` inside the string.
    engine_parser::parse_const_value(default_value).ok()
}

fn directive_location_from_introspection(
    location: cynic_introspection::DirectiveLocation,
) -> registry_v2::DirectiveLocation {
    use cynic_introspection::DirectiveLocation as IntrospectionLocation;
    use registry_v2::DirectiveLocation as EngineLocation;

    match location {
        IntrospectionLocation::Query => EngineLocation::Query,
        IntrospectionLocation::Mutation => EngineLocation::Mutation,
        IntrospectionLocation::Subscription => EngineLocation::Subscription,
        IntrospectionLocation::Field => EngineLocation::Field,
        IntrospectionLocation::FragmentDefinition => EngineLocation::FragmentDefinition,
        IntrospectionLocation::FragmentSpread => EngineLocation::FragmentSpread,
        IntrospectionLocation::InlineFragment => EngineLocation::InlineFragment,
        IntrospectionLocation::VariableDefinition => EngineLocation::VariableDefinition,
        IntrospectionLocation::Schema => EngineLocation::Schema,
        IntrospectionLocation::Scalar => EngineLocation::Scalar,
        IntrospectionLocation::Object => EngineLocation::Object,
        IntrospectionLocation::FieldDefinition => EngineLocation::FieldDefinition,
        IntrospectionLocation::ArgumentDefinition => EngineLocation::ArgumentDefinition,
        IntrospectionLocation::Interface => EngineLocation::Interface,
        IntrospectionLocation::Union => EngineLocation::Union,
        IntrospectionLocation::Enum => EngineLocation::Enum,
        IntrospectionLocation::EnumValue => EngineLocation::EnumValue,
        IntrospectionLocation::InputObject => EngineLocation::InputObject,
        IntrospectionLocation::InputFieldDefinition => EngineLocation::InputFieldDefinition,
    }
}

fn type_from_introspection(ty: cynic_introspection::Type) -> MetaType {
    use cynic_introspection::Type;

    match ty {
        Type::Object(v) => object_from_introspection(v),
        Type::InputObject(v) => input_object_from_introspection(v),
        Type::Enum(v) => enum_from_introspection(v),
        Type::Interface(v) => interface_from_introspection(v),
        Type::Union(v) => union_from_introspection(v),
        Type::Scalar(v) => scalar_from_introspection(v),
    }
}

fn object_from_introspection(object: cynic_introspection::ObjectType) -> MetaType {
    ObjectType::new(object.name, object.fields.into_iter().map(field_from_introspection))
        .with_description(object.description)
        .into()
}

fn deprecated_from_introspection(deprecated: cynic_introspection::Deprecated) -> Deprecation {
    use cynic_introspection::Deprecated as IntrospectionDeprecated;

    match deprecated {
        IntrospectionDeprecated::No => Deprecation::NoDeprecated,
        IntrospectionDeprecated::Yes(reason) => Deprecation::Deprecated { reason },
    }
}

fn input_object_from_introspection(input: cynic_introspection::InputObjectType) -> MetaType {
    engine::registry::InputObjectType::new(input.name, input.fields.into_iter().map(input_value_from_introspection))
        .with_description(input.description)
        .into()
}

fn enum_from_introspection(enum_type: cynic_introspection::EnumType) -> MetaType {
    engine::registry::EnumType::new(
        enum_type.name,
        enum_type.values.into_iter().map(enum_value_from_introspection),
    )
    .with_description(enum_type.description)
    .into()
}

fn enum_value_from_introspection(enum_value: cynic_introspection::EnumValue) -> MetaEnumValue {
    MetaEnumValue {
        name: enum_value.name,
        description: enum_value.description,
        deprecation: deprecated_from_introspection(enum_value.deprecated),
        value: None,
    }
}

fn interface_from_introspection(interface: cynic_introspection::InterfaceType) -> MetaType {
    MetaType::Interface(engine::registry::InterfaceType {
        name: interface.name.clone(),
        description: interface.description,
        fields: interface
            .fields
            .into_iter()
            .map(|v| (v.name.clone(), field_from_introspection(v)))
            .collect(),
        cache_control: Default::default(),
        possible_types: interface.possible_types.into_iter().collect(),
        extends: false,
        rust_typename: interface.name,
    })
}

fn union_from_introspection(union: cynic_introspection::UnionType) -> MetaType {
    MetaType::Union(engine::registry::UnionType {
        name: union.name.clone(),
        description: union.description,
        possible_types: union.possible_types.into_iter().collect(),
        rust_typename: union.name,
        discriminators: None,
    })
}

fn scalar_from_introspection(scalar: cynic_introspection::ScalarType) -> MetaType {
    MetaType::Scalar(engine::registry::ScalarType {
        name: scalar.name,
        description: scalar.description,
        is_valid: None,
        specified_by_url: None,
        parser: engine::registry::ScalarParser::PassThrough,
    })
}

#[cfg(test)]
mod tests {
    use cynic_introspection::query::IntrospectionQuery;

    use crate::conversion::registry_from_introspection;

    #[test]
    fn conversion() {
        let data = include_str!("../tests/swapi_introspection.json");
        let schema = serde_json::from_str::<IntrospectionQuery>(data)
            .unwrap()
            .into_schema()
            .unwrap();

        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!(registry_from_introspection(schema));
        });
    }

    #[test]
    fn array_input_value() {
        let data = include_str!("../tests/countries_introspection.json");
        let schema = serde_json::from_str::<IntrospectionQuery>(data)
            .unwrap()
            .into_schema()
            .unwrap();

        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!(registry_from_introspection(schema));
        });
    }
}
