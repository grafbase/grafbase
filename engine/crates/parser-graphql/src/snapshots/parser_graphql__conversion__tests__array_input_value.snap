---
source: engine/crates/parser-graphql/src/conversion.rs
expression: registry_from_introspection(schema)
---
{
  "types": {
    "Boolean": {
      "S": {
        "n": "Boolean",
        "d": "The `Boolean` scalar type represents `true` or `false`.",
        "p": "PassThrough"
      }
    },
    "Continent": {
      "O": {
        "n": "Continent",
        "f": {
          "code": {
            "n": "code",
            "a": {},
            "t": "ID!"
          },
          "countries": {
            "n": "countries",
            "a": {},
            "t": "[Country!]!"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String!"
          }
        },
        "r": "Continent"
      }
    },
    "ContinentFilterInput": {
      "i": {
        "n": "ContinentFilterInput",
        "i": {
          "code": {
            "n": "code",
            "t": "StringQueryOperatorInput"
          }
        },
        "r": "ContinentFilterInput"
      }
    },
    "Country": {
      "O": {
        "n": "Country",
        "f": {
          "awsRegion": {
            "n": "awsRegion",
            "a": {},
            "t": "String!"
          },
          "capital": {
            "n": "capital",
            "a": {},
            "t": "String"
          },
          "code": {
            "n": "code",
            "a": {},
            "t": "ID!"
          },
          "continent": {
            "n": "continent",
            "a": {},
            "t": "Continent!"
          },
          "currencies": {
            "n": "currencies",
            "a": {},
            "t": "[String!]!"
          },
          "currency": {
            "n": "currency",
            "a": {},
            "t": "String"
          },
          "emoji": {
            "n": "emoji",
            "a": {},
            "t": "String!"
          },
          "emojiU": {
            "n": "emojiU",
            "a": {},
            "t": "String!"
          },
          "languages": {
            "n": "languages",
            "a": {},
            "t": "[Language!]!"
          },
          "name": {
            "n": "name",
            "a": {
              "lang": {
                "n": "lang",
                "t": "String"
              }
            },
            "t": "String!"
          },
          "native": {
            "n": "native",
            "a": {},
            "t": "String!"
          },
          "phone": {
            "n": "phone",
            "a": {},
            "t": "String!"
          },
          "phones": {
            "n": "phones",
            "a": {},
            "t": "[String!]!"
          },
          "states": {
            "n": "states",
            "a": {},
            "t": "[State!]!"
          },
          "subdivisions": {
            "n": "subdivisions",
            "a": {},
            "t": "[Subdivision!]!"
          }
        },
        "r": "Country"
      }
    },
    "CountryFilterInput": {
      "i": {
        "n": "CountryFilterInput",
        "i": {
          "code": {
            "n": "code",
            "t": "StringQueryOperatorInput"
          },
          "continent": {
            "n": "continent",
            "t": "StringQueryOperatorInput"
          },
          "currency": {
            "n": "currency",
            "t": "StringQueryOperatorInput"
          }
        },
        "r": "CountryFilterInput"
      }
    },
    "Float": {
      "S": {
        "n": "Float",
        "d": "The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](https://en.wikipedia.org/wiki/IEEE_floating_point).",
        "p": "PassThrough"
      }
    },
    "ID": {
      "S": {
        "n": "ID",
        "d": "The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `\"4\"`) or integer (such as `4`) input value will be accepted as an ID.",
        "p": "PassThrough"
      }
    },
    "Int": {
      "S": {
        "n": "Int",
        "d": "The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1.",
        "p": "PassThrough"
      }
    },
    "Language": {
      "O": {
        "n": "Language",
        "f": {
          "code": {
            "n": "code",
            "a": {},
            "t": "ID!"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String!"
          },
          "native": {
            "n": "native",
            "a": {},
            "t": "String!"
          },
          "rtl": {
            "n": "rtl",
            "a": {},
            "t": "Boolean!"
          }
        },
        "r": "Language"
      }
    },
    "LanguageFilterInput": {
      "i": {
        "n": "LanguageFilterInput",
        "i": {
          "code": {
            "n": "code",
            "t": "StringQueryOperatorInput"
          }
        },
        "r": "LanguageFilterInput"
      }
    },
    "Query": {
      "O": {
        "n": "Query",
        "f": {
          "continent": {
            "n": "continent",
            "a": {
              "code": {
                "n": "code",
                "t": "ID!"
              }
            },
            "t": "Continent"
          },
          "continents": {
            "n": "continents",
            "a": {
              "filter": {
                "n": "filter",
                "t": "ContinentFilterInput",
                "D": {
                  "__object": {}
                }
              }
            },
            "t": "[Continent!]!"
          },
          "countries": {
            "n": "countries",
            "a": {
              "filter": {
                "n": "filter",
                "t": "CountryFilterInput",
                "D": {
                  "__object": {}
                }
              }
            },
            "t": "[Country!]!"
          },
          "country": {
            "n": "country",
            "a": {
              "code": {
                "n": "code",
                "t": "ID!"
              }
            },
            "t": "Country"
          },
          "language": {
            "n": "language",
            "a": {
              "code": {
                "n": "code",
                "t": "ID!"
              }
            },
            "t": "Language"
          },
          "languages": {
            "n": "languages",
            "a": {
              "filter": {
                "n": "filter",
                "t": "LanguageFilterInput",
                "D": {
                  "__object": {}
                }
              }
            },
            "t": "[Language!]!"
          }
        },
        "r": "Query"
      }
    },
    "State": {
      "O": {
        "n": "State",
        "f": {
          "code": {
            "n": "code",
            "a": {},
            "t": "String"
          },
          "country": {
            "n": "country",
            "a": {},
            "t": "Country!"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String!"
          }
        },
        "r": "State"
      }
    },
    "String": {
      "S": {
        "n": "String",
        "d": "The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
        "p": "PassThrough"
      }
    },
    "StringQueryOperatorInput": {
      "i": {
        "n": "StringQueryOperatorInput",
        "i": {
          "eq": {
            "n": "eq",
            "t": "String"
          },
          "in": {
            "n": "in",
            "t": "[String!]"
          },
          "ne": {
            "n": "ne",
            "t": "String"
          },
          "nin": {
            "n": "nin",
            "t": "[String!]"
          },
          "regex": {
            "n": "regex",
            "t": "String"
          }
        },
        "r": "StringQueryOperatorInput"
      }
    },
    "Subdivision": {
      "O": {
        "n": "Subdivision",
        "f": {
          "code": {
            "n": "code",
            "a": {},
            "t": "ID!"
          },
          "emoji": {
            "n": "emoji",
            "a": {},
            "t": "String"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String!"
          }
        },
        "r": "Subdivision"
      }
    },
    "__Directive": {
      "O": {
        "n": "__Directive",
        "d": "A Directive provides a way to describe alternate runtime execution and type validation behavior in a GraphQL document.\n\nIn some cases, you need to provide options to alter GraphQL's execution behavior in ways field arguments will not suffice, such as conditionally including or skipping a field. Directives provide this by describing additional information to the executor.",
        "f": {
          "args": {
            "n": "args",
            "a": {
              "includeDeprecated": {
                "n": "includeDeprecated",
                "t": "Boolean",
                "D": false
              }
            },
            "t": "[__InputValue!]!"
          },
          "description": {
            "n": "description",
            "a": {},
            "t": "String"
          },
          "isRepeatable": {
            "n": "isRepeatable",
            "a": {},
            "t": "Boolean!"
          },
          "locations": {
            "n": "locations",
            "a": {},
            "t": "[__DirectiveLocation!]!"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String!"
          }
        },
        "r": "__Directive"
      }
    },
    "__DirectiveLocation": {
      "E": {
        "n": "__DirectiveLocation",
        "d": "A Directive can be adjacent to many parts of the GraphQL language, a __DirectiveLocation describes one such possible adjacencies.",
        "e": {
          "ARGUMENT_DEFINITION": {
            "n": "ARGUMENT_DEFINITION",
            "d": "Location adjacent to an argument definition."
          },
          "ENUM": {
            "n": "ENUM",
            "d": "Location adjacent to an enum definition."
          },
          "ENUM_VALUE": {
            "n": "ENUM_VALUE",
            "d": "Location adjacent to an enum value definition."
          },
          "FIELD": {
            "n": "FIELD",
            "d": "Location adjacent to a field."
          },
          "FIELD_DEFINITION": {
            "n": "FIELD_DEFINITION",
            "d": "Location adjacent to a field definition."
          },
          "FRAGMENT_DEFINITION": {
            "n": "FRAGMENT_DEFINITION",
            "d": "Location adjacent to a fragment definition."
          },
          "FRAGMENT_SPREAD": {
            "n": "FRAGMENT_SPREAD",
            "d": "Location adjacent to a fragment spread."
          },
          "INLINE_FRAGMENT": {
            "n": "INLINE_FRAGMENT",
            "d": "Location adjacent to an inline fragment."
          },
          "INPUT_FIELD_DEFINITION": {
            "n": "INPUT_FIELD_DEFINITION",
            "d": "Location adjacent to an input object field definition."
          },
          "INPUT_OBJECT": {
            "n": "INPUT_OBJECT",
            "d": "Location adjacent to an input object type definition."
          },
          "INTERFACE": {
            "n": "INTERFACE",
            "d": "Location adjacent to an interface definition."
          },
          "MUTATION": {
            "n": "MUTATION",
            "d": "Location adjacent to a mutation operation."
          },
          "OBJECT": {
            "n": "OBJECT",
            "d": "Location adjacent to an object type definition."
          },
          "QUERY": {
            "n": "QUERY",
            "d": "Location adjacent to a query operation."
          },
          "SCALAR": {
            "n": "SCALAR",
            "d": "Location adjacent to a scalar definition."
          },
          "SCHEMA": {
            "n": "SCHEMA",
            "d": "Location adjacent to a schema definition."
          },
          "SUBSCRIPTION": {
            "n": "SUBSCRIPTION",
            "d": "Location adjacent to a subscription operation."
          },
          "UNION": {
            "n": "UNION",
            "d": "Location adjacent to a union definition."
          },
          "VARIABLE_DEFINITION": {
            "n": "VARIABLE_DEFINITION",
            "d": "Location adjacent to a variable definition."
          }
        },
        "r": "__DirectiveLocation"
      }
    },
    "__EnumValue": {
      "O": {
        "n": "__EnumValue",
        "d": "One possible value for a given Enum. Enum values are unique values, not a placeholder for a string or numeric value. However an Enum value is returned in a JSON response as a string.",
        "f": {
          "deprecationReason": {
            "n": "deprecationReason",
            "a": {},
            "t": "String"
          },
          "description": {
            "n": "description",
            "a": {},
            "t": "String"
          },
          "isDeprecated": {
            "n": "isDeprecated",
            "a": {},
            "t": "Boolean!"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String!"
          }
        },
        "r": "__EnumValue"
      }
    },
    "__Field": {
      "O": {
        "n": "__Field",
        "d": "Object and Interface types are described by a list of Fields, each of which has a name, potentially a list of arguments, and a return type.",
        "f": {
          "args": {
            "n": "args",
            "a": {
              "includeDeprecated": {
                "n": "includeDeprecated",
                "t": "Boolean",
                "D": false
              }
            },
            "t": "[__InputValue!]!"
          },
          "deprecationReason": {
            "n": "deprecationReason",
            "a": {},
            "t": "String"
          },
          "description": {
            "n": "description",
            "a": {},
            "t": "String"
          },
          "isDeprecated": {
            "n": "isDeprecated",
            "a": {},
            "t": "Boolean!"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String!"
          },
          "type": {
            "n": "type",
            "a": {},
            "t": "__Type!"
          }
        },
        "r": "__Field"
      }
    },
    "__InputValue": {
      "O": {
        "n": "__InputValue",
        "d": "Arguments provided to Fields or Directives and the input fields of an InputObject are represented as Input Values which describe their type and optionally a default value.",
        "f": {
          "defaultValue": {
            "n": "defaultValue",
            "d": "A GraphQL-formatted string representing the default value for this input value.",
            "a": {},
            "t": "String"
          },
          "deprecationReason": {
            "n": "deprecationReason",
            "a": {},
            "t": "String"
          },
          "description": {
            "n": "description",
            "a": {},
            "t": "String"
          },
          "isDeprecated": {
            "n": "isDeprecated",
            "a": {},
            "t": "Boolean!"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String!"
          },
          "type": {
            "n": "type",
            "a": {},
            "t": "__Type!"
          }
        },
        "r": "__InputValue"
      }
    },
    "__Schema": {
      "O": {
        "n": "__Schema",
        "d": "A GraphQL Schema defines the capabilities of a GraphQL server. It exposes all available types and directives on the server, as well as the entry points for query, mutation, and subscription operations.",
        "f": {
          "description": {
            "n": "description",
            "a": {},
            "t": "String"
          },
          "directives": {
            "n": "directives",
            "d": "A list of all directives supported by this server.",
            "a": {},
            "t": "[__Directive!]!"
          },
          "mutationType": {
            "n": "mutationType",
            "d": "If this server supports mutation, the type that mutation operations will be rooted at.",
            "a": {},
            "t": "__Type"
          },
          "queryType": {
            "n": "queryType",
            "d": "The type that query operations will be rooted at.",
            "a": {},
            "t": "__Type!"
          },
          "subscriptionType": {
            "n": "subscriptionType",
            "d": "If this server support subscription, the type that subscription operations will be rooted at.",
            "a": {},
            "t": "__Type"
          },
          "types": {
            "n": "types",
            "d": "A list of all types supported by this server.",
            "a": {},
            "t": "[__Type!]!"
          }
        },
        "r": "__Schema"
      }
    },
    "__Type": {
      "O": {
        "n": "__Type",
        "d": "The fundamental unit of any GraphQL Schema is the type. There are many kinds of types in GraphQL as represented by the `__TypeKind` enum.\n\nDepending on the kind of a type, certain fields describe information about that type. Scalar types provide no information beyond a name, description and optional `specifiedByURL`, while Enum types provide their values. Object and Interface types provide the fields they describe. Abstract types, Union and Interface, provide the Object types possible at runtime. List and NonNull types compose other types.",
        "f": {
          "description": {
            "n": "description",
            "a": {},
            "t": "String"
          },
          "enumValues": {
            "n": "enumValues",
            "a": {
              "includeDeprecated": {
                "n": "includeDeprecated",
                "t": "Boolean",
                "D": false
              }
            },
            "t": "[__EnumValue!]"
          },
          "fields": {
            "n": "fields",
            "a": {
              "includeDeprecated": {
                "n": "includeDeprecated",
                "t": "Boolean",
                "D": false
              }
            },
            "t": "[__Field!]"
          },
          "inputFields": {
            "n": "inputFields",
            "a": {
              "includeDeprecated": {
                "n": "includeDeprecated",
                "t": "Boolean",
                "D": false
              }
            },
            "t": "[__InputValue!]"
          },
          "interfaces": {
            "n": "interfaces",
            "a": {},
            "t": "[__Type!]"
          },
          "kind": {
            "n": "kind",
            "a": {},
            "t": "__TypeKind!"
          },
          "name": {
            "n": "name",
            "a": {},
            "t": "String"
          },
          "ofType": {
            "n": "ofType",
            "a": {},
            "t": "__Type"
          },
          "possibleTypes": {
            "n": "possibleTypes",
            "a": {},
            "t": "[__Type!]"
          },
          "specifiedByURL": {
            "n": "specifiedByURL",
            "a": {},
            "t": "String"
          }
        },
        "r": "__Type"
      }
    },
    "__TypeKind": {
      "E": {
        "n": "__TypeKind",
        "d": "An enum describing what kind of type a given `__Type` is.",
        "e": {
          "ENUM": {
            "n": "ENUM",
            "d": "Indicates this type is an enum. `enumValues` is a valid field."
          },
          "INPUT_OBJECT": {
            "n": "INPUT_OBJECT",
            "d": "Indicates this type is an input object. `inputFields` is a valid field."
          },
          "INTERFACE": {
            "n": "INTERFACE",
            "d": "Indicates this type is an interface. `fields`, `interfaces`, and `possibleTypes` are valid fields."
          },
          "LIST": {
            "n": "LIST",
            "d": "Indicates this type is a list. `ofType` is a valid field."
          },
          "NON_NULL": {
            "n": "NON_NULL",
            "d": "Indicates this type is a non-null. `ofType` is a valid field."
          },
          "OBJECT": {
            "n": "OBJECT",
            "d": "Indicates this type is an object. `fields` and `interfaces` are valid fields."
          },
          "SCALAR": {
            "n": "SCALAR",
            "d": "Indicates this type is a scalar."
          },
          "UNION": {
            "n": "UNION",
            "d": "Indicates this type is a union. `possibleTypes` is a valid field."
          }
        },
        "r": "__TypeKind"
      }
    }
  },
  "directives": {
    "deprecated": {
      "name": "deprecated",
      "description": "Marks an element of a GraphQL schema as no longer supported.",
      "locations": [
        "ArgumentDefinition",
        "EnumValue",
        "FieldDefinition",
        "InputFieldDefinition"
      ],
      "args": {
        "reason": {
          "n": "reason",
          "d": "Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).",
          "t": "String",
          "D": "No longer supported"
        }
      },
      "is_repeatable": false
    },
    "include": {
      "name": "include",
      "description": "Directs the executor to include this field or fragment only when the `if` argument is true.",
      "locations": [
        "Field",
        "FragmentSpread",
        "InlineFragment"
      ],
      "args": {
        "if": {
          "n": "if",
          "d": "Included when true.",
          "t": "Boolean!"
        }
      },
      "is_repeatable": false
    },
    "skip": {
      "name": "skip",
      "description": "Directs the executor to skip this field or fragment when the `if` argument is true.",
      "locations": [
        "Field",
        "FragmentSpread",
        "InlineFragment"
      ],
      "args": {
        "if": {
          "n": "if",
          "d": "Skipped when true.",
          "t": "Boolean!"
        }
      },
      "is_repeatable": false
    },
    "specifiedBy": {
      "name": "specifiedBy",
      "description": "Exposes a URL that specifies the behavior of this scalar.",
      "locations": [
        "Scalar"
      ],
      "args": {
        "url": {
          "n": "url",
          "d": "The URL that specifies the behavior of this scalar.",
          "t": "String!"
        }
      },
      "is_repeatable": false
    }
  },
  "implements": {},
  "query_type": "Query",
  "mutation_type": null,
  "subscription_type": null,
  "disable_introspection": false,
  "enable_federation": false,
  "federation_subscription": false,
  "auth": {
    "allowed_private_ops": 0,
    "allowed_public_ops": 0,
    "allowed_group_ops": {},
    "allowed_owner_ops": 0,
    "provider": null
  },
  "mongodb_configurations": {},
  "http_headers": {},
  "postgres_databases": {},
  "enable_caching": false,
  "enable_partial_caching": false,
  "enable_codegen": false,
  "is_federated": false,
  "operation_limits": {
    "depth": null,
    "height": null,
    "aliases": null,
    "rootFields": null,
    "complexity": null
  },
  "trusted_documents": null,
  "cors_config": null
}
